src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,308 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+                }
+
+                //
+                // Create new blocks. Essentially, we want to transform this:
+                //
+                //   staticBase = helperCall();
+                //
+                // into:
+                //
+                //   if (!isInitialized)
+                //   {
+                //       helperCall(); // we don't use its return value
+                //   }
+                //   staticBase = fastPath;
+                //
+
+                // The initialization check looks like this for JIT:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  AND       int
+                //    |  +--*  IND       int
+                //    |  |  \--*  CNS_INT(h) long   0x.... const ptr
+                //    |  \--*  CNS_INT   int    1 (bit mask)
+                //    \--*  CNS_INT   int    1
+                //
+                // For NativeAOT it's:
+                //
+                // *  JTRUE     void
+                // \--*  EQ        int
+                //    +--*  IND       nint
+                //    |  \--*  ADD       long
+                //    |     +--*  CNS_INT(h) long   0x.... const ptr
+                //    |     \--*  CNS_INT   int    -8 (offset)
+                //    \--*  CNS_INT   int    0
+                //
+                assert(flagAddr.accessType == IAT_VALUE);
+
+                GenTree* cachedStaticBase = nullptr;
+                GenTree* isInitAdrNode;
+                if (isInitOffset != 0)
+                {
+                    assert(IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+                    GenTree* baseAddr = gtNewIconHandleNode((size_t)flagAddr.addr, GTF_ICON_CONST_PTR);
+
+                    // Save it to a temp - we'll be using its value for the replacementNode.
+                    // This leads to some size savings on NativeAOT
+                    if ((staticBaseAddr.addr == flagAddr.addr) && (staticBaseAddr.accessType == flagAddr.accessType))
+                    {
+                        cachedStaticBase = fgInsertCommaFormTemp(&baseAddr);
+                    }
+
+                    // Don't fold ADD(CNS1, CNS2) here since the result won't be reloc-friendly for AOT
+                    isInitAdrNode = gtNewIndir(TYP_I_IMPL, gtNewOperNode(GT_ADD, TYP_I_IMPL, baseAddr,
+                                                                         gtNewIconNode(isInitOffset)));
+                    isInitAdrNode->gtFlags &= ~GTF_EXCEPT;
+                    isInitAdrNode->gtFlags |= (GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
+                }
+                else
+                {
+                    assert(!IsTargetAbi(CORINFO_NATIVEAOT_ABI));
+                    isInitAdrNode = gtNewIndOfIconHandleNode(TYP_INT, (size_t)flagAddr.addr, GTF_ICON_CONST_PTR, true);
+                }
+
+                GenTree* isInitedValue;
+                if (!IsTargetAbi(CORINFO_NATIVEAOT_ABI))
```

Comment:
`isInitOffset != 0` is de-facto same as `if (IsTargetAbi(CORINFO_NATIVEAOT_ABI))` given the asserts above. Would it make look better to just have a single `if (IsTargetAbi(CORINFO_NATIVEAOT_ABI))` check for the whole thing?

Reply:
Addressed